language_service/src/completion.rs

```diff
@@ -1,809 +1,334 @@
-                // more syntax.
-
-                builder.push_locals(compilation, offset, true, true);
-
-                // Item declarations
-                builder.push_item_decl_keywords();
-
-                // Things that go in a block
-                builder.push_stmt_keywords();
-                builder.push_expr_keywords();
-                builder.push_types();
-                builder.push_globals(
-                    compilation,
-                    &context_finder.imports,
-                    insert_open_range,
-                    &context_finder.current_namespace_name,
-                    &indent,
-                );
-
-                // Namespace declarations - least likely to be used, so last
-                builder.push_namespace_keyword();
-            }
-        },
-    };
-
-    CompletionList {
-        items: builder.into_items(),
-    }
-}
-
-fn get_first_non_whitespace_in_source(compilation: &Compilation, package_offset: u32) -> u32 {
-    const QSHARP_MAGIC: &str = "//qsharp";
     let source = compilation
         .user_unit()
         .sources
         .find_by_offset(package_offset)
-        .expect("source should exist in the user source map");
-
-    // Skip the //qsharp magic if it exists (notebook cells)
-    let start = if let Some(qsharp_magic_start) = source.contents.find(QSHARP_MAGIC) {
-        qsharp_magic_start + QSHARP_MAGIC.len()
-    } else {
-        0
-    };
-
-    let source_after_magic = &source.contents[start..];
+        .expect("source should exist");
+    let source_offset: u32 = package_offset - source.offset;
 
-    let first = start
-        + source_after_magic
-            .find(|c: char| !c.is_whitespace())
-            .unwrap_or(source_after_magic.len());
-
-    let first = u32::try_from(first).expect("source length should fit into u32");
-
-    source.offset + first
-}
-
-fn get_indent(compilation: &Compilation, package_offset: u32) -> String {
-    let source = compilation
+    // The parser uses the relative source name to figure out the implicit namespace.
+    let source_name_relative = compilation
         .user_unit()
         .sources
-        .find_by_offset(package_offset)
-        .expect("source should exist in the user source map");
-    let source_offset = (package_offset - source.offset)
-        .try_into()
-        .expect("offset can't be converted to uszie");
-    let before_offset = &source.contents[..source_offset];
-    let mut indent = match before_offset.rfind(|c| c == '{' || c == '\n') {
-        Some(begin) => {
-            let indent = &before_offset[begin..];
-            indent.strip_prefix('{').unwrap_or(indent)
-        }
-        None => before_offset,
-    }
-    .to_string();
-    if !indent.starts_with('\n') {
-        indent.insert(0, '\n');
+        .relative_sources()
+        .find(|s| s.offset == source.offset)
+        .expect("source should exist in the user source map")
+        .name;
```

Comment:
Would it make sense to expose something like `as_relative_source(source: Source)` rather than iterating and cloning all of the sources in the map?
```
let source_name_relative = compilation
        .user_unit()
        .sources
        .as_relative_source(source).name
```

Reply:
Yes, I like that idea.